[BUG] IntegrityError

## Describe the bug

We detected the following errors in our tests
This error occurs when the same token is repeatedly added and removed.

```
/app/ibet-Wallet-API/app/api/v2/admin.py:138: SAWarning: New instance <IDXShareToken at 0x7fa1bf6b40a0> with identity key (<class 'app.model.db.idx_token.IDXShareToken'>, ('0xdec593845c724F1cC49D7671DAdABb5914a14699',), None) conflicts with persistent instance <IDXShareToken at 0x7fa1bf81b490>
  session.commit()
2022-07-14 09:20:40 [FALCON] [ERROR] POST /v2/Admin/Tokens => Traceback (most recent call last):
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1819, in _execute_context
    self.dialect.do_execute(
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 732, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "share_token_pkey"
DETAIL:  Key (token_address)=(0xdec593845c724F1cC49D7671DAdABb5914a14699) already exists.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/falcon/app.py", line 361, in __call__
    responder(req, resp, **params)
  File "/app/ibet-Wallet-API/app/api/v2/admin.py", line 138, in on_post
    session.commit()
  File "<string>", line 2, in commit
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1451, in commit
    self._transaction.commit(_to_root=self.future)
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 829, in commit
    self._prepare_impl()
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 808, in _prepare_impl
    self.session.flush()
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3383, in flush
    self._flush(objects)
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3522, in _flush
    with util.safe_reraise():
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 208, in raise_
    raise exception
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3483, in _flush
    flush_context.execute()
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 456, in execute
    rec.execute(self)
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 630, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 245, in save_obj
    _emit_insert_statements(
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 1097, in _emit_insert_statements
    c = connection._execute_20(
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1631, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 332, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1498, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1862, in _execute_context
    self._handle_dbapi_exception(
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2043, in _handle_dbapi_exception
    util.raise_(
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 208, in raise_
    raise exception
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1819, in _execute_context
    self.dialect.do_execute(
  File "/home/apl/.pyenv/versions/3.10.4/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 732, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "share_token_pkey"
DETAIL:  Key (token_address)=(0xdec593845c724F1cC49D7671DAdABb5914a14699) already exists.
```

--- a/app/api/v2/admin.py
+++ b/app/api/v2/admin.py
@@ -25,11 +25,25 @@ from app import log
 from app import config
 from app.api.common import BaseResource
 from app.contracts import Contract
-from app.errors import InvalidParameterError, DataNotExistsError, AppError, DataConflictError
-from app.model.blockchain import CouponToken, MembershipToken, BondToken, ShareToken
+from app.errors import (
+    InvalidParameterError,
+    DataNotExistsError,
+    AppError,
+    DataConflictError
+)
+from app.model.blockchain import (
+    CouponToken,
+    MembershipToken,
+    BondToken,
+    ShareToken
+)
 from app.model.db import (
     Listing,
-    ExecutableContract
+    ExecutableContract,
+    IDXBondToken,
+    IDXShareToken,
+    IDXMembershipToken,
+    IDXCouponToken
 )
 
 LOG = log.get_logger()
@@ -272,7 +286,7 @@ class Token(BaseResource):
             else token.max_holding_quantity
         max_sell_amount = request_json["max_sell_amount"] if "max_sell_amount" in request_json \
             else token.max_sell_amount
-        owner_address = request_json["owner_address"] if "max_sell_amount" in request_json \
+        owner_address = request_json["owner_address"] if "owner_address" in request_json \
             else token.owner_address
 
         token.is_public = is_public
@@ -333,6 +347,10 @@ class Token(BaseResource):
         try:
             session.query(Listing).filter(Listing.token_address == contract_address).delete()
             session.query(ExecutableContract).filter(ExecutableContract.contract_address == contract_address).delete()
+            session.query(IDXBondToken).filter(IDXBondToken.token_address == contract_address).delete()
+            session.query(IDXShareToken).filter(IDXShareToken.token_address == contract_address).delete()
+            session.query(IDXMembershipToken).filter(IDXMembershipToken.token_address == contract_address).delete()
+            session.query(IDXCouponToken).filter(IDXCouponToken.token_address == contract_address).delete()
         except Exception as err:
             LOG.exception(f"Failed to delete the data: {err}")
             raise AppError()

--- a/tests/app/v2_admin_Token_DELETE_test.py
+++ b/tests/app/v2_admin_Token_DELETE_test.py
@@ -19,7 +19,11 @@ SPDX-License-Identifier: Apache-2.0
 
 from app.model.db import (
     Listing,
-    ExecutableContract
+    ExecutableContract,
+    IDXBondToken,
+    IDXShareToken,
+    IDXMembershipToken,
+    IDXCouponToken
 )
 
 
@@ -28,20 +32,55 @@ class TestAdminTokensDELETE:
     apiurl_base = '/v2/Admin/Token/'
 
     @staticmethod
-    def insert_listing_data(session, _token):
-        token = Listing()
-        token.token_address = _token["token_address"]
-        token.is_public = _token["is_public"]
-        token.max_holding_quantity = _token["max_holding_quantity"]
-        token.max_sell_amount = _token["max_sell_amount"]
-        token.owner_address = _token["owner_address"]
-        session.add(token)
+    def insert_Listing(session, token):
+        listing = Listing()
+        listing.token_address = token["token_address"]
+        listing.is_public = token["is_public"]
+        listing.max_holding_quantity = token["max_holding_quantity"]
+        listing.max_sell_amount = token["max_sell_amount"]
+        listing.owner_address = token["owner_address"]
+        session.add(listing)
+
+    @staticmethod
+    def insert_ExecutableContract(session, token):
+        executable_contract = ExecutableContract()
+        executable_contract.contract_address = token["token_address"]
+        session.add(executable_contract)
+
+    @staticmethod
+    def insert_IDXBondToken(session, token):
+        idx_token = IDXBondToken()
+        idx_token.token_address = token["token_address"]
+        idx_token.token_template = "IbetStraightBond"
+        session.add(idx_token)
+
+    @staticmethod
+    def insert_IDXShareToken(session, token):
+        idx_token = IDXShareToken()
+        idx_token.token_address = token["token_address"]
+        idx_token.token_template = "IbetShare"
+        session.add(idx_token)
+
+    @staticmethod
+    def insert_IDXMembershipToken(session, token):
+        idx_token = IDXMembershipToken()
+        idx_token.token_address = token["token_address"]
+        idx_token.token_template = "IbetMembership"
+        session.add(idx_token)
+
+    @staticmethod
+    def insert_IDXCouponToken(session, token):
+        idx_token = IDXCouponToken()
+        idx_token.token_address = token["token_address"]
+        idx_token.token_template = "IbetCoupon"
+        session.add(idx_token)
 
     ###########################################################################
     # Normal
     ###########################################################################
 
-    # <Normal_1>
+    # Normal_1
+    # Bond
     def test_normal_1(self, client, session):
         token = {
             "token_address": "0x9467ABe171e0da7D6aBDdA23Ba6e6Ec5BE0b4F7b",
@@ -50,16 +89,20 @@ class TestAdminTokensDELETE:
             "max_sell_amount": 50000,
             "owner_address": "0x56f63dc2351BeC560a429f0C646d64Ca718e11D6"
         }
-        self.insert_listing_data(session, token)
+        self.insert_Listing(session, token)
+        self.insert_ExecutableContract(session, token)
+        self.insert_IDXBondToken(session, token)
 
+        # Request target API
         apiurl = self.apiurl_base + token["token_address"]
         resp = client.simulate_delete(apiurl)
 
+        # Assertion
         assert resp.status_code == 200
         assert resp.json['meta'] == {'code': 200, 'message': 'OK'}
 
-        listing = session.query(Listing).\
-            filter(Listing.token_address == token["token_address"]).\
+        listing = session.query(Listing). \
+            filter(Listing.token_address == token["token_address"]). \
             all()
         assert listing == []
 
@@ -67,3 +110,162 @@ class TestAdminTokensDELETE:
             filter(ExecutableContract.contract_address == token["token_address"]). \
             all()
         assert executable_contract == []
+
+        idx_token = session.query(IDXBondToken). \
+            filter(IDXBondToken.token_address == token["token_address"]). \
+            all()
+        assert idx_token == []
+
+    # Normal_2
+    # Share
+    def test_normal_2(self, client, session):
+        token = {
+            "token_address": "0x9467ABe171e0da7D6aBDdA23Ba6e6Ec5BE0b4F7b",
+            "is_public": True,
+            "max_holding_quantity": 100,
+            "max_sell_amount": 50000,
+            "owner_address": "0x56f63dc2351BeC560a429f0C646d64Ca718e11D6"
+        }
+        self.insert_Listing(session, token)
+        self.insert_ExecutableContract(session, token)
+        self.insert_IDXShareToken(session, token)
+
+        # Request target API
+        apiurl = self.apiurl_base + token["token_address"]
+        resp = client.simulate_delete(apiurl)
+
+        # Assertion
+        assert resp.status_code == 200
+        assert resp.json['meta'] == {'code': 200, 'message': 'OK'}
+
+        listing = session.query(Listing). \
+            filter(Listing.token_address == token["token_address"]). \
+            all()
+        assert listing == []
+
+        executable_contract = session.query(ExecutableContract). \
+            filter(ExecutableContract.contract_address == token["token_address"]). \
+            all()
+        assert executable_contract == []
+
+        idx_token = session.query(IDXShareToken). \
+            filter(IDXShareToken.token_address == token["token_address"]). \
+            all()
+        assert idx_token == []
+
+    # Normal_3
+    # Membership
+    def test_normal_3(self, client, session):
+        token = {
+            "token_address": "0x9467ABe171e0da7D6aBDdA23Ba6e6Ec5BE0b4F7b",
+            "is_public": True,
+            "max_holding_quantity": 100,
+            "max_sell_amount": 50000,
+            "owner_address": "0x56f63dc2351BeC560a429f0C646d64Ca718e11D6"
+        }
+        self.insert_Listing(session, token)
+        self.insert_ExecutableContract(session, token)
+        self.insert_IDXMembershipToken(session, token)
+
+        # Request target API
+        apiurl = self.apiurl_base + token["token_address"]
+        resp = client.simulate_delete(apiurl)
+
+        # Assertion
+        assert resp.status_code == 200
+        assert resp.json['meta'] == {'code': 200, 'message': 'OK'}
+
+        listing = session.query(Listing). \
+            filter(Listing.token_address == token["token_address"]). \
+            all()
+        assert listing == []
+
+        executable_contract = session.query(ExecutableContract). \
+            filter(ExecutableContract.contract_address == token["token_address"]). \
+            all()
+        assert executable_contract == []
+
+        idx_token = session.query(IDXMembershipToken). \
+            filter(IDXMembershipToken.token_address == token["token_address"]). \
+            all()
+        assert idx_token == []
+
+    # Normal_4
+    # Coupon
+    def test_normal_4(self, client, session):
+        token = {
+            "token_address": "0x9467ABe171e0da7D6aBDdA23Ba6e6Ec5BE0b4F7b",
+            "is_public": True,
+            "max_holding_quantity": 100,
+            "max_sell_amount": 50000,
+            "owner_address": "0x56f63dc2351BeC560a429f0C646d64Ca718e11D6"
+        }
+        self.insert_Listing(session, token)
+        self.insert_ExecutableContract(session, token)
+        self.insert_IDXCouponToken(session, token)
+
+        # Request target API
+        apiurl = self.apiurl_base + token["token_address"]
+        resp = client.simulate_delete(apiurl)
+
+        # Assertion
+        assert resp.status_code == 200
+        assert resp.json['meta'] == {'code': 200, 'message': 'OK'}
+
+        listing = session.query(Listing). \
+            filter(Listing.token_address == token["token_address"]). \
+            all()
+        assert listing == []
+
+        executable_contract = session.query(ExecutableContract). \
+            filter(ExecutableContract.contract_address == token["token_address"]). \
+            all()
+        assert executable_contract == []
+
+        idx_token = session.query(IDXCouponToken). \
+            filter(IDXCouponToken.token_address == token["token_address"]). \
+            all()
+        assert idx_token == []
+
+    # Normal_5
+    # Multiple token type(template)
+    def test_normal_5(self, client, session):
+        token = {
+            "token_address": "0x9467ABe171e0da7D6aBDdA23Ba6e6Ec5BE0b4F7b",
+            "is_public": True,
+            "max_holding_quantity": 100,
+            "max_sell_amount": 50000,
+            "owner_address": "0x56f63dc2351BeC560a429f0C646d64Ca718e11D6"
+        }
+        self.insert_Listing(session, token)
+        self.insert_ExecutableContract(session, token)
+        self.insert_IDXBondToken(session, token)
+        self.insert_IDXCouponToken(session, token)
+
+        # Request target API
+        apiurl = self.apiurl_base + token["token_address"]
+        resp = client.simulate_delete(apiurl)
+
+        # Assertion
+        assert resp.status_code == 200
+        assert resp.json['meta'] == {'code': 200, 'message': 'OK'}
+
+        listing = session.query(Listing). \
+            filter(Listing.token_address == token["token_address"]). \
+            all()
+        assert listing == []
+
+        executable_contract = session.query(ExecutableContract). \
+            filter(ExecutableContract.contract_address == token["token_address"]). \
+            all()
+        assert executable_contract == []
+
+        idx_token = session.query(IDXBondToken). \
+            filter(IDXBondToken.token_address == token["token_address"]). \
+            all()
+        assert idx_token == []
+
+        idx_token = session.query(IDXCouponToken). \
+            filter(IDXCouponToken.token_address == token["token_address"]). \
+            all()
+        assert idx_token == []

--- a/tests/app/v2_admin_Token_POST_test.py
+++ b/tests/app/v2_admin_Token_POST_test.py
@@ -85,7 +85,6 @@ class TestAdminTokenPOST:
         assert listing.max_sell_amount == request_params["max_sell_amount"]
         assert listing.owner_address == request_params["owner_address"]
 
-
     ###########################################################################
     # Error
     ###########################################################################

